skimage/restoration/_denoise.py

```diff
@@ -146,112 +146,41 @@ def _denoise_tv_chambolle_3d(im, weight=100, eps=2.e-4, n_iter_max=200):
 
     """
 
-    px = np.zeros_like(im)
-    py = np.zeros_like(im)
-    pz = np.zeros_like(im)
-    gx = np.zeros_like(im)
-    gy = np.zeros_like(im)
-    gz = np.zeros_like(im)
+    ndim = im.ndim
+    p = np.zeros(im.shape + (im.ndim, ), dtype=im.dtype)
+    g = np.zeros_like(p)
     d = np.zeros_like(im)
     i = 0
     while i < n_iter_max:
-        d = - px - py - pz
-        d[1:] += px[:-1]
-        d[:, 1:] += py[:, :-1]
-        d[:, :, 1:] += pz[:, :, :-1]
-
-        out = im + d
-        E = (d ** 2).sum()
-
-        gx[:-1] = np.diff(out, axis=0)
-        gy[:, :-1] = np.diff(out, axis=1)
-        gz[:, :, :-1] = np.diff(out, axis=2)
-        norm = np.sqrt(gx ** 2 + gy ** 2 + gz ** 2)
-        E += weight * norm.sum()
-        norm *= 0.5 / weight
-        norm += 1.
-        px -= 1. / 6. * gx
-        px /= norm
-        py -= 1. / 6. * gy
-        py /= norm
-        pz -= 1 / 6. * gz
-        pz /= norm
-        E /= float(im.size)
-        if i == 0:
-            E_init = E
-            E_previous = E
+        if i > 0:
+            d = -p.sum(-1)
+            slices_d = [slice(None), ] * ndim
+            slices_p = [slice(None), ] * (ndim + 1)
+            for ax in range(ndim):
+                slices_d[ax] = slice(1, None)
+                slices_p[ax] = slice(0, -1)
+                slices_p[-1] = ax
```

Comment:
Did you check on large images (say 1000x1000) the performance of the new nD function compared to the exisiting 2D and 3D functions? I wonder whether it wouldn't be better to have the first dimension of p (instead of the last one) corresponding to the dimensions of ``im``, since operations would then be performed on more contiguous memory chunks.